uavcan/node/62805.Heartbeat.1.0.uavcan

```diff
@@ -0,0 +1,47 @@
+#
+# Abstract node status information.
+#
+# All UAVCAN nodes are required to publish this message periodically.
+# This is the only data type that must be supported by all nodes.
+#
+# The default subject ID 62805 is 1111010101010101 in binary. The alternating bit pattern at the end
+# helps transceiver synchronization (e.g., on CAN-based networks) and on some transports permits
+# automatic bit rate detection (e.g., CAN bus).
+#
+
+# Publication period must not exceed this limit.
+# The period should not change as long as the node is running.
+uint16 MAX_PUBLICATION_PERIOD = 1   # [second]
+
+# If a node fails to publish this message in this amount of time, it should be considered offline.
+uint16 OFFLINE_TIMEOUT = 3          # [second]
+
+# The uptime seconds counter should never overflow. The counter will reach the upper limit in ~136 years,
+# upon which time it should stay at 0xFFFFFFFF until the node is restarted.
+#
+# Other nodes may detect that a remote node has restarted when this value jumps backwards.
+uint32 uptime                       # [second]
+
+# Abstract node health.
+uint2 HEALTH_STATUS_OK       = 0    # Nominal. Color: none or green.
+uint2 HEALTH_STATUS_ADVISORY = 1    # Minor parameter deviation or minor issue. Color: blue/cyan.
+uint2 HEALTH_STATUS_CAUTION  = 2    # Major problem or failure. Color: amber/yellow.
+uint2 HEALTH_STATUS_WARNING  = 3    # Fault, loss of function. Color: red.
+truncated uint2 health_status
+
+# The current operating mode.
+#
+# The mode OFFLINE can be used for informing other network participants that the sending node has ceased its
+# activities or about to do so. In this case, other nodes will not have to wait for the OFFLINE_TIMEOUT to
+# expire before detecting that the sending node is no longer available.
+#
+# Reserved values can be used in future revisions of the specification.
+uint3 MODE_OPERATIONAL      = 0     # Normal operating mode.
+uint3 MODE_INITIALIZATION   = 1     # Initialization is in progress; this mode is entered immediately after startup.
+uint3 MODE_MAINTENANCE      = 2     # E.g. calibration, self-test, etc.
+uint3 MODE_SOFTWARE_UPDATE  = 3     # New software/firmware is being loaded or the bootloader is running.
+uint3 MODE_OFFLINE          = 7     # The node is no longer available.
+truncated uint3 mode
+
+# Optional, vendor-specific node status code, e.g. a fault code or a status bitmask.
+truncated uint19 vendor_specific_status_code
```

Comment:
I guess `node_specific_data` is a better name? As the same vendor might use it differently for different products?

Can we use one of the bits to make this field optional?

```python
truncated uint18[<=1] node_specific_data
```

Alternatively, If we wanted to leverage the CAN-FD capabilities without messing things up for CAN2.0B we could perhaps do the following:

NodeSpecificStatusCode:
```python
@union 
uint64[0] none # Is this a specification compliant way of creating a zero sized variant? We should probably allow `void0` or`()` to avoid this.
uint8[2] data_short # (alternatively `uint16 data_short`) The shorter alternative for vendor specific data.
uint8[8] data_long # (alternatively `uint64 data_long`) The longer alternative for vendor specific data. Highly recommended to avoid unless it's known that the underlying protocol will be able to send the whole frame in a single transfer.
# We could also have a `uint8[<=255] string` variant if we wanted to go yolo, of course urging users not to send strings that will cause a multi frame transfer.
```

---

I think this flexibility would be ok as this code/data is used for diagnosis/logging/debuging and not read by other nodes. A tool could for instance plot 64 bits and gray out 48 of them if the short variant was used. Or gray out all of them if the none variant was used.

Reply:
Displaying this information is not the only major use case. The other one is other nodes checking the vendor-specific status code automatically, e.g. for extended application-specific self-diagnostics. Unions are flexible and fun, but they really add a lot of extra complexity and extend the testing matrix, so I would only use them when absolutely necessary. Vote for static typing.

![image](https://user-images.githubusercontent.com/3298404/47171260-e7ff3600-d310-11e8-985e-fe620eff2b63.png)